rust-maven-plugin/src/main/java/io/questdb/maven/rust/Crate.java

```diff
@@ -0,0 +1,505 @@
+        final Process process = processBuilder.start();
+        Executors.newSingleThreadExecutor().submit(() -> {
+            new BufferedReader(new InputStreamReader(process.getInputStream()))
+                    .lines()
+                    .forEach(log::info);
+        });
+
+        final int exitCode = process.waitFor();
+        if (exitCode != 0) {
+            throw new MojoExecutionException(
+                    "Cargo command failed with exit code " + exitCode);
+        }
+    }
+
+    private void cargo(List<String> args) throws MojoExecutionException, MojoFailureException {
+        String cargoPath = getCargoPath();
+        final List<String> cmd = new ArrayList<>();
+        cmd.add(cargoPath);
+        cmd.addAll(args);
+        log.info("Working directory: " + crateRoot);
+        if (!params.environmentVariables.isEmpty()) {
+            log.info("Environment variables:");
+            for (String key : params.environmentVariables.keySet()) {
+                log.info("  " + key + "=" + Shlex.quote(
+                        params.environmentVariables.get(key)));
+            }
+        }
+        log.info("Running: " + Shlex.quote(cmd));
+        try {
+            runCommand(cmd);
+        } catch (IOException | InterruptedException e) {
+            CargoInstalledChecker.INSTANCE.check(log, cargoPath);
+            throw new MojoFailureException("Failed to invoke cargo", e);
+        }
+    }
+
+    private void addCargoArgs(List<String> args) {
+        args.add("--target-dir");
+        args.add(targetDir.toAbsolutePath().toString());
+
+        if (params.release) {
+            args.add("--release");
+        }
+
+        if (params.allFeatures) {
+            args.add("--all-features");
+        }
+
+        if (params.noDefaultFeatures) {
+            args.add("--no-default-features");
+        }
+
+        if (params.features != null && params.features.length > 0) {
+            args.add("--features");
+            args.add(String.join(",", params.features));
+        }
+
+        if (params.tests) {
+            args.add("--tests");
+        }
+
+        if (params.extraArgs != null) {
+            Collections.addAll(args, params.extraArgs);
+        }
+    }
+
+    public void build() throws MojoExecutionException, MojoFailureException {
+        List<String> args = new ArrayList<>();
+        args.add("build");
+        addCargoArgs(args);
+        cargo(args);
+    }
+
+    public void test() throws MojoExecutionException, MojoFailureException {
+        List<String> args = new ArrayList<>();
+        args.add("test");
+        addCargoArgs(args);
+        cargo(args);
+    }
+
+    private Path resolveCopyToDir() throws MojoExecutionException {
+
+        Path copyToDir = params.copyToDir;
+
+        if (copyToDir == null) {
+            return null;
+        }
+
+        if (params.copyWithPlatformDir) {
+            final String osName = System.getProperty("os.name").toLowerCase();
```

Comment:
Yes.

I was thinking it would be cleaner (from a dependencies point of view) if the plugin didn't use the `jar-jni` library.

Reply:
I originally thought that it would be better to copy and paste some of this code around with the goal of reducing dependencies.
You're right and I've changed it so that `rust-maven-plugin` repends on `jar-jni`.

Done.